AmplifyPlugins/API/AWSAPICategoryPlugin/AWSAPIPlugin+Reachability.swift

```diff
@@ -17,7 +17,16 @@ extension AWSAPIPlugin {
 
     @available(iOS 13.0, *)
     public func reachabilityPublisher(for apiName: String?) throws -> AnyPublisher<ReachabilityUpdate, Never>? {
-        let hostName = try determineHostName(apiName: apiName)
+        let endpoint = try pluginConfig.endpoints.getConfig(for: apiName)
+        guard let hostName = endpoint.baseURL.host else {
+            let error = APIError.invalidConfiguration("Invalid endpoint configuration",
+                """
+                baseURL does not contain a valid hostname
+                """
+            )
+            throw error
+        }
+
         if let networkReachability = reachabilityMap[hostName] {
```

Comment:
I don't see where `reachabilityMap` is declared, but is this thread safe? What happens if different threads invoke this method at the same time?

Reply:
nice catch, reachability is declared [here](https://github.com/aws-amplify/amplify-ios/blob/main/AmplifyPlugins/API/AWSAPICategoryPlugin/AWSAPIPlugin.swift#L43-L58) with an underlying `iReachabilityMap` to work around having `@available(iOS 13.0, *)` on a stored property. (we can have a computed property, but not stored property with `@available`)

I just tested this with a `DispatchQueue.concurrentPerform` and is not thread safe. The code looks like it is retrieving the underlying map, and setting it, so when two threads try to access the dictionary, i was able to reproduce a data race

updated PR to perform the work on reachabilityMap under a NSLock, i think the cleanest solution so far. my attempt to convert it to AtomicValue had data races in various places due the access performed in the computed property: https://github.com/aws-amplify/amplify-ios/pull/1200